### PR TITLE
#2796  Made Url a layout

### PR DESCRIPTION
**Pull request by Bjego**
Hey guys,

the issue is now older than one year. Here is a tiny update to handle variables in the url for the webservice target and it keeps the existing behaviour. 

With best regards

### CODE REVIEW

File: src/NLog/Targets/WebServiceTarget.cs [hunk 1]
```diff
@@ -122,7 +123,20 @@ public WebServiceTarget(string name) : this()
         /// Gets or sets the web service URL.
         /// </summary>
         /// <docgen category='Web Service Options' order='10' />
-        public Uri Url { get; set; }
+        /// 
+        public Layout Url { get; set; }
+        internal Uri _renderedUrl
```
Comment (snakefoot) [inline src/NLog/Targets/WebServiceTarget.cs]: This is a propery not a private member. Please remove the underscore.

Comment (snakefoot) [reply src/NLog/Targets/WebServiceTarget.cs]: Why having both Url and _renderedUrl ?

File: src/NLog/Targets/WebServiceTarget.cs [hunk 2]
```diff
@@ -122,7 +123,20 @@ public WebServiceTarget(string name) : this()
...
+        {
+            get
+            {
+                var uri = RenderLogEvent(Url, LogEventInfo.CreateNullEvent());
```
Comment (snakefoot) [inline src/NLog/Targets/WebServiceTarget.cs]: No need to render every time when using `LogEventInfo.CreateNullEvent()`. Instead cache the rendering, and reset the cache in InitializeTarget.

**Comment by snakefoot** [comment]
Thank you for the PR. The reason for not resolving the issue is because it is a breaking change, and requires NLog 5.0. Like this PR is also a breaking change, and cannot be included in NLog 4.7

Luckily NLog 5.0 is under way, but right now focus is on rewiring the logic for better interaction with external dependency injection systems (Ex. Microsoft Dependency Injection).

**Comment by Bjego** [comment]
Okay - I've written my own custom target. I guess that is good enough for now. And I'll have a look at the official target again, when Nlog 5.0 is ready

**Comment by snakefoot** [comment]
You might be interested in: https://nuget.org/NLog.Targets.Http


Sent from my Sony Xperia


---- Bjego wrote ----


Okay - I've written my own custom target. I guess that is good enough for now. And I'll have a look at the official target again, when Nlog 5.0 is ready

**Comment by Bjego** [comment]
@snakefoot no that target isn't good as well. It's also using strings instead of layouts...
I just need a simple splunk target - and that should be able to read the configs from the app settings
my custom target looks like this now.
```csharp
using NLog;
using NLog.Common;
using NLog.Layouts;
using NLog.Targets;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace CRM2.NlogTargets
{
  [Target("Splunk")]
  /// <summary>
  /// https://github.com/NLog/NLog/wiki/How-to-write-a-custom-async-target
  /// </summary>
  public class SplunkTarget : AsyncTaskTarget
  {
    /// <summary>
    /// https://github.com/NLog/NLog/wiki/How-to-write-a-custom-async-target
    /// </summary>
    public SplunkTarget()
    {
    }
    protected override void Dispose(bool disposing)
    {
      if (disposing)
      {
        _client.Dispose();
        _client = null;
      }
      base.Dispose(disposing);
    }
    private static SemaphoreSlim _semaphoreSlim;
    private string _url;
    private HttpClient _client;
    public Layout Url { get; set; }
    public Layout Token { get; set; }
    public bool OldSsl { get; set; }


    protected override async Task WriteAsyncTask(IList<LogEventInfo> logEvents, CancellationToken cancellationToken)
    {
      var localClient = _client ?? await InitHttpClient();

      var bodies = from e in logEvents select Layout.Render(e);
      var bodyToSend = string.Join(" ", bodies);

      InternalLogger.Trace($"Sending to Splunk ({logEvents.Count} Events): {bodyToSend}");
      var splunksResponse = await localClient.PostAsync(_url, new StringContent(bodyToSend, Encoding.UTF8, "application/json"), cancellationToken);
      splunksResponse.EnsureSuccessStatusCode();
    }

    private async Task<HttpClient> InitHttpClient()
    {
      try
      {
        await _semaphoreSlim.WaitAsync();
        if (_client == null)
        {
          var token = Token.Render(LogEventInfo.CreateNullEvent());
          InternalLogger.Info($"Using Splunk-Token: {token}");
          _client = new HttpClient();
          _client.DefaultRequestHeaders.Add("Authorization", $"Splunk {token}");
        }
        if (string.IsNullOrWhiteSpace(_url))
        {
          var splunkUrl = $"{Url.Render(LogEventInfo.CreateNullEvent())}/services/collector/event";
          InternalLogger.Info($"Using Splunk-Url:{splunkUrl}");
          _url = splunkUrl;
        }
        return _client;
      }
      finally
      {
        _semaphoreSlim.Release();
      }
    }

    protected override void InitializeTarget()
    {
      _semaphoreSlim = new SemaphoreSlim(1, 1);
      if (OldSsl)
      {
        ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls | SecurityProtocolType.Tls11 | SecurityProtocolType.Tls12 | SecurityProtocolType.Ssl3;
      }
      base.InitializeTarget();

    }

    protected override Task WriteAsyncTask(LogEventInfo logEvent, CancellationToken cancellationToken)
    {
      throw new NotImplementedException();
    }
  }
}

```

and the nlog config is:
```xml
 <target xsi:type="Splunk"
         name="splunk"
         Url="${configsetting:name=Splunk.Server}"
         Token="${configsetting:name=Splunk.Token}"
         OldSsl="true"
         BatchSize="2000"
         TaskDelayMilliseconds="1000"
         OverflowAction="Grow"
         RetryCount="20"
         RetryDelayMilliseconds="2500" >
      <layout type='JsonLayout'>
        <attribute name='sourcetype' layout='_json' />
        <attribute name="event" encode="false">
          <layout xsi:type="JsonLayout">
            <attribute name="Host" layout="${machinename}" />
            <attribute name="App" layout="${appName}"/>
            <attribute name="Exception" encode="false" layout="${exception:format=@}" />
            <attribute name="Message" layout="${message}" />
            <attribute name="Thread" layout="${threadid}" />
            <attribute name="TraceIdentifier" layout="${aspnet-TraceIdentifier}"/>
            <attribute name="Logger" layout="${logger}" />
            <attribute name="Environment" layout="${aspnet-environment}"/>
            <attribute name="Level" layout="${level}"/>
            <attribute name="Date" layout="${longdate}"/>
            <attribute name="Controller" layout="${aspnet-mvc-controller}"/>
            <attribute name="Action" layout="${aspnet-mvc-action}"/>
            <attribute name="User" layout="${aspnet-user-identity}"/>
            <attribute name="Client Ip" layout="${aspnet-request-ip}"/>
            <attribute name="aspnet-request-url" layout="${aspnet-request-url}"/>
            <attribute name="aspnet-request-method" layout="${aspnet-request-method}"/>
            <attribute name="aspnet-request-referrer" layout="${aspnet-request-referrer}"/>
            <attribute name="aspnet-request-form" layout="${aspnet-request-form}"/>
            <attribute name="aspnet-request-host" layout="${aspnet-request-host}"/>
            <attribute name="aspnet-request-useragent" layout="${aspnet-request-useragent}"/>
            <attribute name="aspnet-request-querystring " layout="${aspnet-request-querystring}" />
            <attribute name="aspnet-request-posted-body" layout="${aspnet-request-posted-body}" />
            <attribute name="aspnet-request-contenttype" layout="${aspnet-request-contenttype}" />
            <attribute name="aspnet-user-authtype" layout="${aspnet-user-authtype}"/>
          </layout>
        </attribute>
      </layout>
    </target>
```

Is it worth it to share it in the official Nlog repo? Not sure could create a PR, if it's helpful

**Comment by snakefoot** [comment]
@Bjego You don't need _semaphoreSlim, as AsyncTaskTarget will automatically wait for task to complete before starting new.

But happy to see that AsyncTaskTarget is working out for you. Making me consider to make a HttpClientTarget that inherits from AsyncTaskTarget.

**Comment by Bjego** [comment]
@snakefoot your idea sounds good - maybe this could go into one of the next releases of Nlog? And you maybe deprecate the Webservice Target (for Json).. 
Please keep in mind that the Url and Headers (for authentication) should be layouts :). Makes it easier to stage an app between - dev, UAT and Prod.

**Comment by Bjego** [comment]
@snakefoot I think a "Splunk" / "HttpClientTarget"- Target doesn't need more than the settings I've provided and looks like a simple implementation.
Plus you wouldn't need any external vendors to write the plugin - updating and fixing would be easier for everyone, because Nlog is a good maintained project, instead of those "lone wolfs", which provide specific extentions.
And in Nlog 5 you could inject the HttpClient via the Ihttpclient factory.. If a user wants to inject his/her own httpClient.

**Pull request closed without merging**